### PR TITLE
New versioning patches

### PR DESCRIPTION
**Pull request by rafatower**
hey @jgoizueta can you please review? mostly documentation and the branch is not final as I did not get to execute the pysal tests yet.

cc/ @ohasselblad please account for the production environment limitations. The target of this extension is the user's production database.

### CODE REVIEW

**Comment by jgoizueta** [comment]
did you forget to add a requirements.txt ? or are we using the setup.py install_requires ?

**Comment by jgoizueta** [comment]
Under servers provisioned with PG 9.5 (at least this is happening in my local env and in one production server) scipy build fails:
```
  Running setup.py install for scipy ... error
    Complete output from command /home/ubuntu/crankshaft/myenv/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-XQ6f1Q/scipy/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-X2roXj-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/ubuntu/crankshaft/myenv/include/site/python2.7/scipy:
    lapack_opt_info:
    openblas_lapack_info:
      libraries openblas not found in ['/home/ubuntu/crankshaft/myenv/lib', '/usr/local/lib', '/usr/lib', '/usr/lib/x86_64-linux-gnu']
      NOT AVAILABLE
    
    lapack_mkl_info:
    mkl_info:
      libraries mkl,vml,guide not found in ['/home/ubuntu/crankshaft/myenv/lib', '/usr/local/lib', '/usr/lib', '/usr/lib/x86_64-linux-gnu']
      NOT AVAILABLE
    
      NOT AVAILABLE
    
    atlas_3_10_threads_info:
    Setting PTATLAS=ATLAS
      libraries tatlas,tatlas not found in /home/ubuntu/crankshaft/myenv/lib
      libraries lapack_atlas not found in /home/ubuntu/crankshaft/myenv/lib
      libraries tatlas,tatlas not found in /usr/local/lib
      libraries lapack_atlas not found in /usr/local/lib
      libraries tatlas,tatlas not found in /usr/lib
      libraries lapack_atlas not found in /usr/lib
      libraries tatlas,tatlas not found in /usr/lib/x86_64-linux-gnu
      libraries lapack_atlas not found in /usr/lib/x86_64-linux-gnu
    <class 'numpy.distutils.system_info.atlas_3_10_threads_info'>
      NOT AVAILABLE
    
    atlas_3_10_info:
      libraries satlas,satlas not found in /home/ubuntu/crankshaft/myenv/lib
      libraries lapack_atlas not found in /home/ubuntu/crankshaft/myenv/lib
      libraries satlas,satlas not found in /usr/local/lib
      libraries lapack_atlas not found in /usr/local/lib
      libraries satlas,satlas not found in /usr/lib
      libraries lapack_atlas not found in /usr/lib
      libraries satlas,satlas not found in /usr/lib/x86_64-linux-gnu
      libraries lapack_atlas not found in /usr/lib/x86_64-linux-gnu
    <class 'numpy.distutils.system_info.atlas_3_10_info'>
      NOT AVAILABLE
    
    atlas_threads_info:
    Setting PTATLAS=ATLAS
      libraries ptf77blas,ptcblas,atlas not found in /home/ubuntu/crankshaft/myenv/lib
      libraries lapack_atlas not found in /home/ubuntu/crankshaft/myenv/lib
      libraries ptf77blas,ptcblas,atlas not found in /usr/local/lib
      libraries lapack_atlas not found in /usr/local/lib
      libraries ptf77blas,ptcblas,atlas not found in /usr/lib
      libraries lapack_atlas not found in /usr/lib
      libraries ptf77blas,ptcblas,atlas not found in /usr/lib/x86_64-linux-gnu
      libraries lapack_atlas not found in /usr/lib/x86_64-linux-gnu
    <class 'numpy.distutils.system_info.atlas_threads_info'>
      NOT AVAILABLE
    
    atlas_info:
      libraries f77blas,cblas,atlas not found in /home/ubuntu/crankshaft/myenv/lib
      libraries lapack_atlas not found in /home/ubuntu/crankshaft/myenv/lib
      libraries f77blas,cblas,atlas not found in /usr/local/lib
      libraries lapack_atlas not found in /usr/local/lib
      libraries f77blas,cblas,atlas not found in /usr/lib
      libraries lapack_atlas not found in /usr/lib
      libraries f77blas,cblas,atlas not found in /usr/lib/x86_64-linux-gnu
      libraries lapack_atlas not found in /usr/lib/x86_64-linux-gnu
    <class 'numpy.distutils.system_info.atlas_info'>
      NOT AVAILABLE
    
    /home/ubuntu/crankshaft/myenv/local/lib/python2.7/site-packages/numpy/distutils/system_info.py:1552: UserWarning:
        Atlas (http://math-atlas.sourceforge.net/) libraries not found.
        Directories to search for the libraries can be specified in the
        numpy/distutils/site.cfg file (section [atlas]) or by setting
        the ATLAS environment variable.
      warnings.warn(AtlasNotFoundError.__doc__)
    lapack_info:
      libraries lapack not found in ['/home/ubuntu/crankshaft/myenv/lib', '/usr/local/lib', '/usr/lib', '/usr/lib/x86_64-linux-gnu']
      NOT AVAILABLE
    
    /home/ubuntu/crankshaft/myenv/local/lib/python2.7/site-packages/numpy/distutils/system_info.py:1563: UserWarning:
        Lapack (http://www.netlib.org/lapack/) libraries not found.
        Directories to search for the libraries can be specified in the
        numpy/distutils/site.cfg file (section [lapack]) or by setting
        the LAPACK environment variable.
      warnings.warn(LapackNotFoundError.__doc__)
    lapack_src_info:
      NOT AVAILABLE
    
    /home/ubuntu/crankshaft/myenv/local/lib/python2.7/site-packages/numpy/distutils/system_info.py:1566: UserWarning:
        Lapack (http://www.netlib.org/lapack/) sources not found.
        Directories to search for the sources can be specified in the
        numpy/distutils/site.cfg file (section [lapack_src]) or by setting
        the LAPACK_SRC environment variable.
      warnings.warn(LapackSrcNotFoundError.__doc__)
      NOT AVAILABLE
    
    Running from scipy source directory.
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-XQ6f1Q/scipy/setup.py", line 265, in <module>
        setup_package()
      File "/tmp/pip-build-XQ6f1Q/scipy/setup.py", line 262, in setup_package
        setup(**metadata)
      File "/home/ubuntu/crankshaft/myenv/local/lib/python2.7/site-packages/numpy/distutils/core.py", line 135, in setup
        config = configuration()
      File "/tmp/pip-build-XQ6f1Q/scipy/setup.py", line 182, in configuration
        config.add_subpackage('scipy')
      File "/home/ubuntu/crankshaft/myenv/local/lib/python2.7/site-packages/numpy/distutils/misc_util.py", line 1002, in add_subpackage
        caller_level = 2)
      File "/home/ubuntu/crankshaft/myenv/local/lib/python2.7/site-packages/numpy/distutils/misc_util.py", line 971, in get_subpackage
        caller_level = caller_level + 1)
      File "/home/ubuntu/crankshaft/myenv/local/lib/python2.7/site-packages/numpy/distutils/misc_util.py", line 908, in _get_configuration_from_setup_py
        config = setup_module.configuration(*args)
      File "scipy/setup.py", line 15, in configuration
        config.add_subpackage('linalg')
      File "/home/ubuntu/crankshaft/myenv/local/lib/python2.7/site-packages/numpy/distutils/misc_util.py", line 1002, in add_subpackage
        caller_level = 2)
      File "/home/ubuntu/crankshaft/myenv/local/lib/python2.7/site-packages/numpy/distutils/misc_util.py", line 971, in get_subpackage
        caller_level = caller_level + 1)
      File "/home/ubuntu/crankshaft/myenv/local/lib/python2.7/site-packages/numpy/distutils/misc_util.py", line 908, in _get_configuration_from_setup_py
        config = setup_module.configuration(*args)
      File "scipy/linalg/setup.py", line 20, in configuration
        raise NotFoundError('no lapack/blas resources found')
    numpy.distutils.system_info.NotFoundError: no lapack/blas resources found
```

**Comment by rafatower** [comment]
I did not add the `requirements.txt` as I couldn't get it to pass the tests. My concern is building from a solid foundation and this branch doesn't achieve it. For the record:
```
======================================================================
FAIL: test_G_Local_Binary (pysal.esda.tests.test_getisord.G_Local_Tester)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/pysal/pysal/esda/tests/test_getisord.py", line 33, in test_G_Local_Binary
    self.assertAlmostEquals(lg.Zs[0], -1.0136729, places=7)
AssertionError: -2.3379874918175281 != -1.0136729 within 7 places

======================================================================
FAIL: test_G_Local_Row_Standardized (pysal.esda.tests.test_getisord.G_Local_Tester)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/pysal/pysal/esda/tests/test_getisord.py", line 38, in test_G_Local_Row_Standardized
    self.assertAlmostEquals(lg.Zs[0], -0.62074534, places=7)
AssertionError: -0.95447939666040116 != -0.62074534 within 7 places

...

======================================================================
FAIL: test_error_3eq (test_sur_error.Test_SUR_error)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/pysal/pysal/spreg/tests/test_sur_error.py", line 107, in test_error_3eq
    [  1.194989e-001,   1.174869e+001,   7.172248e-032]])},RTOL)
  File "/tmp/pysal/pysal/spreg/tests/test_sur_error.py", line 12, in dict_compare
    np.testing.assert_allclose(actual[i],desired[i],rtol)
  File "/home/ubuntu/www/crankshaft/src/py/dev2/local/lib/python2.7/site-packages/numpy/testing/utils.py", line 1359, in assert_allclose
    verbose=verbose, header=header)
  File "/home/ubuntu/www/crankshaft/src/py/dev2/local/lib/python2.7/site-packages/numpy/testing/utils.py", line 713, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-05, atol=0

(mismatch 22.2222222222%)
 x: array([[  1.345558e-001,   3.325677e+001,   1.629451e-242],
       [  1.205318e-001,   1.816984e+001,   8.945122e-074],
       [  1.092657e-001,   5.409796e+000,   6.309664e-008]])
 y: array([[  1.345557e-001,   3.325679e+001,   1.628656e-242],
       [  1.205317e-001,   1.816985e+001,   8.943986e-074],
       [  1.092657e-001,   5.409796e+000,   6.309653e-008]])

----------------------------------------------------------------------
Ran 498 tests in 135.152s

FAILED (failures=11)
```

I'm also testing with 9.5. This should be independent of the server version (9.3 or 9.5). Let's focus on 9.5 as it is the future. Consider 9.3 legacy. FDW do not work properly with 9.3 BTW.

**Comment by rafatower** [comment]
@javisantana just FYI: we're having some trouble getting the python modules to work. This is pretty much regardless of whether we install them in a virtual env or not and even using them from plpython.

**Comment by ohasselblad** [comment]
Hey @rafatower and @jgoizueta, do you need anything from @stuartlynn or me to move this forward? Now that the Data Observatory is going, him and I are pivoting to focus on crankshaft more.

**Comment by rafatower** [comment]
We'd need to fit this into the roadmap and I don't think there's room for it in the next sprint. cc/ @javisantana

**Comment by stuartlynn** [comment]
Hey all. No rush on looking at this just wanted to record it in the discussion somewhere. I know you guys are swamped with other things in this sprint

I have been playing with crankshaft on envisage, trying to get a few libraries working to do density contours. scikit-image, the module that I needed to grab the contours themselves required a newer version of numpy. After playing around a bit I managed to get numpy 1.11.0 working from within crankshaft using the dev environment and it seems to play nice with the system. 

I know we have restrictions on what we can install based on the system requirements but couldn't we upgrade to numpy 1.11.0 in the dev environment? It seems to work pretty well. 

screen shot 2016-05-20 at 3 54 38 pm

I upgraded numpy and installed the other required libraries manually using pip install but it should be possible to add these to the setup.py file as well.

**Comment by rafatower** [comment]
We need to work on the requirements for crankshaft. When we first thought
of this extension many things were different. It cannot have a dependency
upon observatory. We no longer plan to connect all the tables in the
observatory through FDW. The deploy process makes the observatory DB
volatile... to sum up: it cannot be a placeholder of "stuff" with unbound
libraries if we plan to have it for the editor and serve all our users.

I'll ping you to work some requirements out so that we can proceed.

On Mon, May 23, 2016 at 4:04 PM, Stuart Lynn <notifications@github.com>
wrote:

> Hey all. No rush on looking at this just wanted to record it in the
> discussion somewhere. I know you guys are swamped with other things in this
> sprint
>
> I have been playing with crankshaft on envisage, trying to get a few
> libraries working to do density contours. scikit-image, the module that I
> needed to grab the contours themselves required a newer version of numpy.
> After playing around a bit I managed to get numpy 1.11.0 working from
> within crankshaft using the dev environment and it seems to play nice with
> the system.
>
> I know we have restrictions on what we can install based on the system
> requirements but couldn't we upgrade to numpy 1.11.0 in the dev
> environment? It seems to work pretty well.
>
> [image: screen shot 2016-05-20 at 3 54 38 pm]
> <https://cloud.githubusercontent.com/assets/209838/15473259/2938b95a-20cd-11e6-93a7-d0d7c9288099.png>
>
> I upgraded numpy and installed the other required libraries manually using
> pip install but it should be possible to add these to the setup.py file as
> well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/CartoDB/crankshaft/pull/9#issuecomment-220989420>
>



-- 
*Rafa de la Torre*
DataServices Tech Lead www.cartodb.com

**Comment by ohasselblad** [comment]
Having no dependency on data observatory is fine with us. Looking forward to working on the requirements.

**Comment by ohasselblad** [comment]
If there's anything we can read to better understand the limitations of the production environment, we'd love to read it.

**Comment by andrewxhill** [comment]
@javisantana this is the one mentioning Numpy, but it isn't a dedicated ticket

File: src/py/crankshaft/setup.py [hunk 1]
```diff
@@ -40,7 +40,11 @@
 
     # The choice of component versions is dictated by what's
     # provisioned in the production servers.
-    install_requires=['pysal==1.11.0','numpy==1.6.1','scipy==0.17.0'],
+    install_requires=[
+        'numpy>=1.10.4,<2',
```
Comment (javisantana) [inline src/py/crankshaft/setup.py]: @luisbosque we talk about this next meeting

Comment (luisbosque) [inline src/py/crankshaft/setup.py]: :+1:

**Pull request closed without merging**